### ISSUE TITLE
`cascade_invocations` false positives

### ISSUE THREAD

**Issue by vr19860507** [dart-lang/linter#1821]
cascade_invocations false positives

```
void main() {
  final List<int> items = [];
  var g = 7;
  items.clear();
  items.addAll([g]); // LINT
}
```

**Comment by jamesderlin** [comment]
Why is this a false positive?  What's wrong with:

```dart
items
  ..clear()
  ..addAll([g]);
```
?

**Comment by vr19860507** [comment]
> Why is this a false positive? What's wrong with:
> 
> ```dart
> items
>   ..clear()
>   ..addAll([g]);
> ```
> 
> ?

The problem is with void!

**Comment by bwilkerson** [comment]
Could you be a little more explicit? I don't see any particular problem caused by `void` in this example.

**Comment by jamesderlin** [comment]
@vr19860507 If you believe there would be some problem from `clear()` returning `void`, the point of using the cascade operator (`..`) instead of normal member access (`.`) is that the return value is ignored when chaining to the next method call.

**Comment by vr19860507** [comment]
@jamesderlin @bwilkerson 

Just enable "cascade_invocations" And just paste this code
```
void main() {
  final List<int> items = [];
  var g = 7;
  items.clear();
  items.addAll([g]); // LINT
}
```
into the editor.

And you will understand what I'm talking about!

**Comment by jamesderlin** [comment]
@vr19860507 The lint is telling you that you can use the cascade operator instead, as I showed in https://github.com/dart-lang/linter/issues/1821#issuecomment-673202687.  The lint seems correct; what makes you think that it is a *false* positive?

**Comment by bwilkerson** [comment]
Try running the following code:
```
void main() {
  final List<int> items = [];
  var g = 7;

  items.clear();
  items.addAll([g]);
  print(items);

  items..clear()..addAll([g]);
  print(items);
}
```
This will print out two identical lines because the two ways of writing the code are equivalent. In both cases any exiting contents of the list are removed (the invocation of `clear`) and then a single item is added to the list (the invocation of `addAll`).

I suspect that what wasn't clear is that the cascade operator means to invoke both `clear` and `addAll` on the result of the expression before the first cascade operator, which in this case is `items`.

**Comment by vr19860507** [comment]
@bwilkerson @jamesderlin Checked it now, it works!
```
items
  ..clear()
  ..addAll([g]);
```
But I remember exactly that when I opened this issue it did not work... I don't know, maybe I was mistaken, or maybe they fixed it...

Thank you anyway!

**Comment by bwilkerson** [comment]
You're welcome. I'm glad it's working for you.
